[Blogpost] Configurable Automation for OpenSearch ML Use Cases (#2698)
opensearch-project/project-website: _posts/2024-04-05-Configurable-Automation-for-OpenSearch-ML-Use-Cases.md

---
layout: post
title: "Configurable automation for OpenSearch ML use cases"
authors:
    - kazabdu
    - amitgalitz
    - dwiddis
    - jpalis
    - hnyng
    - ohltyler
    - minalsha
date: 2024-04-05
categories:
  - technical-posts
meta_keywords: Flow Framework, OpenSearch plugins, Machine Learning
meta_description: Explore the simplicity of integrating Machine Learning capabilities within OpenSearch through an innovative and groundbreaking framework designed to simplify complex setup tasks.
---

> In OpenSearch, to use machine learning (ML) offerings, such as semantic, hybrid, and multimodal search, you often have to grapple with complex setup and preprocessing tasks. Additionally, you must write verbose queries, which can be a time-consuming and error-prone process.

Reviewer: This sentence has a lot of commas. I tried to rewrite it to make it better but couldn't really do much better. So I guess it's fine as is! :|

Author: @kolchfa-aws any inputs here?

> In this blog post, we introduce the OpenSearch Flow Framework plugin, [released in version 2.13](https://opensearch.org/blog/2.13-is-ready-for-download/) and designed to streamline this cumbersome process. By using this plugin, you can simplify complex setups with just one click. We've provided automated templates, enabling you to create connectors, register models, deploy them, and register agents and tools through a single API call. This eliminates the complexity of calling multiple APIs and orchestrating setups based on the responses.

Reviewer: Nit: "click" is kind of GUI centric and we're still API.  Can we maybe say "one simple API call?"

## Before the Flow Framework plugin

> Previously, setting up semantic search involves *4 separate API* calls outlined in the [semantic search documentation](https://opensearch.org/docs/latest/search-plugins/semantic-search/):

Reviewer: Put the closing `*` after "calls"

---
layout: post
title: "Configurable automation for OpenSearch ML use cases"
authors:
    - kazabdu
    - amitgalitz
    - dwiddis
    - jpalis
    - hnyng
    - ohltyler
    - minalsha
date: 2024-04-05
categories:
  - technical-posts
meta_keywords: Flow Framework, OpenSearch plugins, Machine Learning
meta_description: Explore the simplicity of integrating Machine Learning capabilities within OpenSearch through an innovative and groundbreaking framework designed to simplify complex setup tasks.
---

In OpenSearch, to use ML offerings, such as semantic, hybrid, and multimodal search, you often have to grapple with complex setup and preprocessing tasks. Additionally, you must write verbose queries, which can be time-consuming and error-prone.

> In this blog post, we introduce the OpenSearch Flow Framework plugin, released in version 2.13, designed to streamline this cumbersome process. By using this plugin, you can simplify complex setups with just one click. We've provided automated templates, enabling you to create connectors, register models, deploy them, and register agents and tools through a single API call. This eliminates the complexity of calling multiple APIs and orchestrating setups based on the responses.

Reviewer 2: Please add a link here: "released in version 2.13" to the release blog or release notes. 
https://github.com/opensearch-project/opensearch-build/blob/main/release-notes/opensearch-release-notes-2.13.0.md

## Before the Flow Framework plugin

Previously, setting up semantic search involved the steps outlined in the [semantic search documentation](https://opensearch.org/docs/latest/search-plugins/semantic-search/):

1. Create a connector for a remote model, specifying pre- and post-processing functions.
1. Register an embedding model using the connector ID obtained in the previous step.
1. Configure an ingest pipeline to generate vector embeddings using the model ID of the registered model.

> 1. Create a k-NN index and add the pipeline created in the previous step.

Reviewer: It's not clear to the reader how complex this is, particularly since the same steps are essentially repeated on steps 38-40.  We need to highlight in the sentence above (line 25) that these require 4 separate API calls, perhaps adding the words "copy and paste" when refer to "using the X ID"

> This complex setup typically required you to be familiar with the OpenSearch ML Commons APIs. However, we are simplifying this experience through the Flow Framework plugin. Let's demonstrate how the plugin simplifies this process using the preceding semantic search example.

Reviewer: Don't think we need "typically"

## With the Flow Framework plugin

In this example, you will configure the `semantic_search_with_cohere_embedding_query_enricher` workflow template. The workflow created using this template performs the following configuration steps:

* Deploys an externally hosted Cohere model
* Creates an ingest pipeline using the model
* Creates a sample k-NN index and configures a search pipeline to define the default model ID for that index

### Step 1: Create and provision the workflow

Using the `semantic_search_with_cohere_embedding_query_enricher` workflow template, you provision the workflow with just one required field---the API key for the Cohere Embed model:

```json
POST /_plugins/_flow_framework/workflow?use_case=semantic_search_with_cohere_embedding_query_enricher&provision=true
{
    "create_connector.credential.key" : "<YOUR API KEY>"
}
```

OpenSearch responds with a unique workflow ID, simplifying the tracking and management of the setup process:

```json
{
  "workflow_id" : "8xL8bowB8y25Tqfenm50"
}
```

Note: The workflow in the previous step creates a default k-NN index. The default index name is `my-nlp-index`:

```json
{

>   "create_index.name": "my-nlp-index"

Reviewer: His is confusing having this separated from the line 47 API, particularly with the response JSON in between.  I *think* this goes in the block under line 49, but even I'm not sure.   Make it clear, perhaps including it as a line 50 after describing that it's optional, or perhaps repeating the whole REST call with both lines.

Author: I will remove the json completely

}
```

You can customize the template default values by providing the new values in the request body. For a comprehensive list of default parameter values for this workflow template, see [Cohere Embed semantic search defaults](https://github.com/opensearch-project/flow-framework/blob/2.13/src/main/resources/defaults/cohere-embedding-semantic-search-defaults.json).

### Step 2: Ingest documents into the index

Once the workflow is provisioned, you can ingest documents into the index created by the workflow:

```json

> PUT /my-nlp-index/_doc/1

Reviewer: We should probably use the path without the document ID included.  Having the `1` here and the `s1` in the "id" field is confusing.  (We can leave out the "id" field as well since it's not the same as the "_id".)

{
  "passage_text": "Hello world",
  "id": "s1"
}
```

### Step 3: Perform vector search

Performing a vector search on the index is equally straightforward. Using a neural query clause, you can easily retrieve relevant results:

```json
GET /my-nlp-index/_search
{
  "_source": {
    "excludes": [
      "passage_embedding"
    ]
  },
  "query": {
    "neural": {
      "passage_embedding": {
        "query_text": "Hi world",

>         "k": 100

Reviewer: Maybe a smaller k?

Reviewer: I meant a number less than 100, heh.  Like `"k": 10`

Author: Done

      }
    }
  }
}
```

With Flow Framework, we've simplified this complex setup process, enabling you to focus on your tasks without the burden of navigating complex APIs. Our goal is for you to use OpenSearch seamlessly, unlocking new possibilities in your projects.

## Viewing workflow resources

The workflow you created provisioned all the necessary resources for semantic search. To view the provisioned resources, call the Get Workflow Status API and provide the `workflowID` for your workflow:

```
GET /_plugins/_flow_framework/workflow/8xL8bowB8y25Tqfenm50/_status
```

## Additional default use cases

> You can explore more default use cases by viewing [substitution templates](https://github.com/opensearch-project/flow-framework/tree/2.13/src/main/resources/substitutionTemplates) with their corresponding [defaults](https://github.com/opensearch-project/flow-framework/tree/2.13/src/main/resources/defaults).

Reviewer 3: ```suggestion
You can explore more default use cases by viewing [substitution templates](https://github.com/opensearch-project/flow-framework/tree/2.13/src/main/resources/substitutionTemplates) and their corresponding [defaults](https://github.com/opensearch-project/flow-framework/tree/2.13/src/main/resources/defaults).
```

## Creating custom use cases

> You can tailor templates according to your requirements. For more information, see [sample templates](https://github.com/opensearch-project/flow-framework/tree/main/sample-templates) and the [documentation](https://opensearch.org/docs/latest/automating-configurations/index/).

Reviewer 4: ```suggestion
You can tailor templates according to your requirements. For more information, see [these sample templates](https://github.com/opensearch-project/flow-framework/tree/main/sample-templates) and the [Automating configurations](https://opensearch.org/docs/latest/automating-configurations/index/) documentation.
```

## Next steps

> In our ongoing efforts to enhance user experience and streamline the process of provisioning the ML offerings of OpenSearch, we have an exciting plan in the pipeline. We aim to develop a user-friendly drag-and-drop frontend interface. This interface will simplify complex steps involved in provisioning ML features, thereby allowing you to seamlessly configure and deploy your workflows. Stay tuned for updates on this exciting development!

Reviewer 4: ```suggestion
In our ongoing efforts to enhance the user experience and streamline the process of provisioning OpenSearch ML offerings, we have some exciting plans on our roadmap. We aim to develop a user-friendly drag-and-drop frontend interface. This interface will simplify the complex steps involved in provisioning ML features, thereby allowing you to seamlessly configure and deploy your workflows. Stay tuned for updates on this exciting development!
```

> If you have any comments or suggestions, please comment on the RFCs:

Reviewer 4: ```suggestion
If you have any comments or suggestions, you can comment on the following RFCs:
```

## Before the Flow Framework plugin

> Previously, setting up semantic search involves *4 separate API* calls outlined in the [semantic search documentation](https://opensearch.org/docs/latest/search-plugins/semantic-search/):

Reviewer 4: "involved four". Comma after "calls".

}
```

> Note: The workflow in the previous step creates a default k-NN index. The default index name is `my-nlp-index`

Reviewer 4: Add terminating period.
